test.py

```diff
@@ -640,9 +640,14 @@ def adjust_suite_name(name):
             suites = root.findall('.//TestSuite')
             for suite in suites:
                 suite.attrib['name'] = adjust_suite_name(suite.attrib['name'])
-                skipped = suite.findall('./TestCase[@reason="disabled"]')
-                for e in skipped:
-                    suite.remove(e)
+                for test in suite.findall('.//TestCase'):
+                    if test.get('reason') == 'disabled':
+                        suite.remove(test)
+                        continue
+                    test_name = test.attrib['name']
+                    # append id at end of the name, to differentiate different runs of
+                    # the same test, in case --repeat is greater than 1
+                    test.attrib['name'] = f'{test_name}.{self.id}'
```

Comment:
Why do you think it's a good idea to have an own id for each run, rather than just add the failure to the existing node?

Reply:
I am not sure what you mean by “add the failure to the existing node”. how can we assume there is a failure here?

Reply:
Do you meant to suggest adding the failure if any to the test cases with the same name of the same test suite? If that's the case, how can we teach Jenkins that "this is the failure"? By naming it differently?

Reply:
JUnit format has a schema. In other words, it's a set of predefined fields that Jenkins understand. If there is a test under the same name that succeeded, you can add failures: section to it, that would mean that this test has actually failed.

Reply:
https://www.ibm.com/docs/en/developer-for-zos/14.1?topic=formats-junit-xml-format

Reply:
the above is informal description, this is a formal xsd apparently https://github.com/windyroad/JUnit-Schema/blob/master/JUnit.xsd

Reply:
Please note this XML is not JUnit. But the boost’s own XML log format.

what I don’t understand is, how your suggestion addresses the problem I want to address in this change.

Reply:
IIUC, your proposal is more complicated and misleading — to find all tests with the same name and mark them as failure if any of them fails.

Reply:
> Please note this XML is not JUnit. But the boost’s own XML log format.

Why do you think so? The routine is called write_consolidated_junit_xml(). Maybe we should discuss this outside this pull request, because either one of us doesn't understand what this code is doing or another doesn't understand the pull request? I suggest we continue on slack.
> 
> what I don’t understand is, how your suggestion addresses the problem I want to address in this change.

Here's what I think is happening. If you have --repeat, write_consolidate_junix_xml() wil create duplicate entires in xml output. Jenkins will pick a random entry and will use it for output. What we need to do is to avoid duplicates. There are two ways of avoiding duplicates. Give each repeat a unique name in the output (bloats the output, but I would perhaps accept it if it was achieved but a bit of less of python kung-fu) or just append each run to the failures section of the test case xml. I suggested the second. And no, it is not *all* tests with the same name, the output is supposed to already contain only unique names.

Reply:
> > Please note this XML is not JUnit. But the boost’s own XML log format.
> 
> Why do you think so? The routine is called write_consolidated_junit_xml(). 

i think so because, we run the tests with the command line arguments like, `--logger=HRF,test_suite:XML,test_suite,<path_to_xunit_xml_file>`. see https://github.com/scylladb/scylladb/blob/27f7cc403281015c1fcecb8ad5c4db8fd9d70b00/test.py#L609 . and per boost's document, the argument to the `--logger` is a list separated by ":", so what we are passing to this option are two loggers:

1. `HRF,test_suite`: this is the human readable format, which is what we are reading in the stdout
2. `XML,test_suite,<path_to_xunit_xml_file>`: this is the machine readable XML format. in our case, it is "XML" but not "JUNIT". which are documented at https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/test_output/log_formats/log_xml_format.html and https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/test_output/log_formats/log_junit_format.html respectively.

also, FWIW, we parse the output of the unit tests at https://github.com/scylladb/scylla-pkg/blob/f613d0c7703fd85085481cc7f81ac9b4a9a98ba1/scripts/jenkins-pipelines/groovy_modules/test.groovy#L64-L66 using both xUnit and JUnit . but what actually does the heavy lifting is XUnit, see https://github.com/scylladb/scylla-pkg/blob/f613d0c7703fd85085481cc7f81ac9b4a9a98ba1/scripts/jenkins-pipelines/groovy_modules/test.groovy#L719-L727 . as we are using Jenkins's [xUnit plugin](https://plugins.jenkins.io/xunit/)'s builtin tool to [translate](https://github.com/jenkinsci/xunit-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/xunit/types/boosttest-1.5.0.xsd) boost's own format to JUnit,  so it can support boost::test's own XML logger.

> Maybe we should discuss this outside this pull request, because either one of us doesn't understand what this code is doing or another doesn't understand the pull request? I suggest we continue on slack.

i just put some more background above. will ping you once you are online.

> 
> > what I don’t understand is, how your suggestion addresses the problem I want to address in this change.
> 
> Here's what I think is happening. If you have --repeat, write_consolidate_junix_xml() wil create duplicate entires in xml output. Jenkins will pick a random entry and will use it for output. What we need to do is to avoid duplicates. There are two ways of avoiding duplicates. Give each repeat a unique name in the output (bloats the output, but I would perhaps accept it if it was achieved but a bit of less of python kung-fu) 

bloats? how come it is worse that what we have?

> or just append each run to the failures section of the test case xml. I suggested the second. And no, it is not _all_ tests with the same name, the output is supposed to already contain only unique names.